homeassistant/components/homekit_controller/humidifier.py

```diff
@@ -0,0 +1,224 @@
+            ATTR_MAX_HUMIDITY: self.max_humidity,
+            ATTR_AVAILABLE_MODES: self.available_modes,
+        }
+
+        return data
+
+    @property
+    def state_attributes(self) -> Dict[str, Any]:
+        """Return the optional state attributes."""
+        data = {
+            ATTR_MODE: self.mode,
+            ATTR_HUMIDITY: self.target_humidity,
+            ATTR_CURRENT_HUMIDITY: self.service.value(
+                CharacteristicsTypes.RELATIVE_HUMIDITY_CURRENT
+            ),
+        }
+
+        dehumidifier_threshold = self.service.value(
+            CharacteristicsTypes.RELATIVE_HUMIDITY_DEHUMIDIFIER_THRESHOLD
+        )
+        if dehumidifier_threshold is not None:
+            data[ATTR_DEHUMIDIFIER_THRESHOLD] = dehumidifier_threshold
+
+        humidifier_threshold = self.service.value(
+            CharacteristicsTypes.RELATIVE_HUMIDITY_HUMIDIFIER_THRESHOLD
+        )
+        if humidifier_threshold is not None:
+            data[ATTR_HUMIDIFIER_THRESHOLD] = humidifier_threshold
+
+        return data
+
+    @property
+    def target_humidity(self) -> Optional[int]:
+        """Return the humidity we try to reach."""
+        return self.service.value(self._char_threshold_for_current_mode())
+
+    @property
+    def mode(self) -> Optional[str]:
+        """Return the current mode, e.g., home, auto, baby.
+
+        Requires SUPPORT_MODES.
+        """
+        mode = self.service.value(
+            CharacteristicsTypes.CURRENT_HUMIDIFIER_DEHUMIDIFIER_STATE
+        )
+        return HK_MODE_TO_HA.get(mode, "unknown")
+
+    @property
+    def available_modes(self) -> Optional[List[str]]:
+        """Return a list of available modes.
+
+        Requires SUPPORT_MODES.
+        """
+        available_modes = [
+            "off",
+            "auto",
+        ]
+        humidifier_threshold = self.service.value(
+            CharacteristicsTypes.RELATIVE_HUMIDITY_HUMIDIFIER_THRESHOLD
+        )
+        if humidifier_threshold is not None:
+            available_modes.append("humidifying")
+
+        dehumidifier_threshold = self.service.value(
+            CharacteristicsTypes.RELATIVE_HUMIDITY_DEHUMIDIFIER_THRESHOLD
+        )
+        if dehumidifier_threshold is not None:
+            available_modes.append("dehumidifying")
+
+        return available_modes
+
+    async def async_set_humidity(self, humidity: int) -> None:
+        """Set new target humidity."""
+        await self.async_put_characteristics(
+            {self._char_threshold_for_current_mode(): humidity}
+        )
+
+    async def async_set_mode(self, mode: str) -> None:
+        """Set new mode."""
+
+        if mode == "off":
+            await self.async_put_characteristics({CharacteristicsTypes.ACTIVE: False})
+        else:
+            new_mode = HA_MODE_TO_HK.get(mode, "unknown")
+            await self.async_put_characteristics(
+                {CharacteristicsTypes.TARGET_HUMIDIFIER_DEHUMIDIFIER_STATE: new_mode}
```

Comment:
Should we set `CharacteristicsTypes.ACTIVE: True` every time ? Or working like this ?

Reply:
When setting the mode to off in HomeKit, it also sets the ACTIVE characteristic to off. So these two are managed in tandem in HomeKit. When turning active to on, it will restore the mode automatically to whatever it was set to previously (auto, humidifying, dehumidifying). Additionally, there is not a target mode for off. 

In other words, it would be redundant.

Reply:
Confirmed.